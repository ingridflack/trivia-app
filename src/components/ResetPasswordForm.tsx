import {
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  useToast,
} from "@chakra-ui/react";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../services/authService";

import { useNavigate, useParams } from "react-router-dom";

interface ResetPasswordFormValues {
  password: string;
  passwordConfirmation: string;
}

export default function ResetPasswordForm() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<ResetPasswordFormValues>();
  const toast = useToast();
  const navigate = useNavigate();
  const { token } = useParams();

  const onSubmit: SubmitHandler<ResetPasswordFormValues> = async (values) => {
    if (!token) return;

    try {
      await AuthService.resetPassword(token, values);

      toast({
        title: "Email sent.",
        description: "Check your inbox.",
        status: "success",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });

      navigate("/login");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  if (!token) {
    return (
      <div>
        <p>Invalid token.</p>
      </div>
    );
  }

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      style={{
        display: "flex",
        flexDirection: "column",
      }}
    >
      <FormControl isInvalid={!!errors.password}>
        <FormLabel
          marginTop="16px"
          htmlFor="password"
          fontSize="small"
          marginBottom="4px"
        >
          Password
        </FormLabel>
        <Input
          type="password"
          placeholder="Password"
          id="password"
          {...register("password", {
            required: "Password is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
          })}
        />
        <FormErrorMessage>
          {errors.password && errors.password.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.passwordConfirmation}>
        <FormLabel
          marginTop="16px"
          htmlFor="passwordConfirmation"
          fontSize="small"
          marginBottom="4px"
        >
          Password confirmation
        </FormLabel>
        <Input
          type="password"
          placeholder="Password Confirmation"
          id="passwordConfirmation"
          {...register("passwordConfirmation", {
            required: "Password confirmation is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
            validate: {
              isEqual: (value, { password }) =>
                value !== password ? "The passwords do not match" : true,
            },
          })}
        />
        <FormErrorMessage>
          {errors.passwordConfirmation && errors.passwordConfirmation.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        colorScheme="purple"
        size="lg"
        width="100%"
        maxWidth="400px"
        marginTop="16px"
        isLoading={isSubmitting}
        type="submit"
      >
        Reset
      </Button>
    </form>
  );
}
