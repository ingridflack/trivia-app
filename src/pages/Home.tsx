import { Box, Button, Container, Image, Text } from "@chakra-ui/react";
import Header from "../components/Header";
import { Link } from "react-router-dom";

export default function Home() {
  return (
    <>
      <Header />
      <Box width="100%" backgroundColor="#AE79F8" overflow="hidden">
        <Container paddingY="80px" position="relative">
          <Text
            fontSize="6xl"
            lineHeight="1.2"
            fontWeight="bold"
            color="white"
            maxWidth="550"
          >
            Challenge yourself in a trivia match!
          </Text>

          <Button
            backgroundColor="green.400"
            color="white"
            width="180px"
            _hover="green.400"
            marginTop="38px"
            as={Link}
            to="/trivia/config"
          >
            Start a trivia
          </Button>
          <Image
            src="/src/assets/home.svg"
            alt="Home image"
            position="absolute"
            right="-160px"
            bottom="-60px"
            width="674px"
            height="418px"
            pointerEvents="none"
          />
        </Container>
      </Box>
    </>
  );
}
