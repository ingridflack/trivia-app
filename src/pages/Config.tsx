import { Box, Card, Container, Image, Text } from "@chakra-ui/react";
import TriviaConfig from "../components/TriviaConfig";
import Header from "../components/Header";
import { Footer } from "../components/Footer";

export default function Config() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/src/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
