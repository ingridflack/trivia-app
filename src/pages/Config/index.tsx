import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import TriviaConfig from "../../components/TriviaConfig";
import Header from "../../components/Header";
import { Footer } from "../../components/Footer";
import * as UserService from "../../services/userService";
import * as TriviaService from "../../services/triviaService";
import { SubmitHandler, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useState } from "react";
import { MultiValue } from "react-select";
import { UserSearchSelectOption } from "../../types/sharedTypes";
import useAuth from "../../hooks/useAuth";

export interface GameConfigValues {
  category: string;
  difficulty: string;
  amount: number;
  gameMode: string;
  type: string;
}

export default function Config() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<GameConfigValues>({
    defaultValues: {
      difficulty: "any",
      type: "any",
      amount: 5,
    },
  });

  const [gameMode, setGameMode] = useState("single");
  const navigate = useNavigate();
  const toast = useToast();
  const [invitedUsers, setInvitedUsers] = useState<
    MultiValue<UserSearchSelectOption>
  >([]);
  const { userData } = useAuth();

  const onSubmit: SubmitHandler<GameConfigValues> = async (values) => {
    try {
      const invitedUsersIds =
        gameMode === "multi"
          ? invitedUsers.map((option) => option.value)
          : undefined;

      const body = {
        ...values,
        invitedUsers: invitedUsersIds,
      };

      const { data } = await TriviaService.createTrivia(body);

      navigate(`/trivia/${data.triviaId}`);
    } catch {
      console.log("error");

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  const handleGameModeChange = (value: string) => {
    setGameMode(value);
  };

  const handleInviteUser = (value: MultiValue<UserSearchSelectOption>) => {
    setInvitedUsers(value);
  };

  const handleLoadUsers = async (username: string) => {
    const currentUserId = userData ? userData.id : "";

    const { data } = await UserService.search({ username });
    return data
      .filter((user) => user._id !== currentUserId)
      .map((user) => ({ value: user._id, label: user.username }));
  };

  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig
              onLoadUsers={handleLoadUsers}
              onInviteUser={handleInviteUser}
              onGameModeChange={handleGameModeChange}
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              register={register}
              isSubmitting={isSubmitting}
              gameMode={gameMode}
              invitedUsers={invitedUsers}
            />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
