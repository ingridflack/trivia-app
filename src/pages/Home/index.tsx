import {
  Box,
  Button,
  Container,
  Heading,
  Image,
  Skeleton,
  Stack,
  Text,
} from "@chakra-ui/react";
import Header from "../../components/Header";
import { Link, useNavigate } from "react-router-dom";
import { PendingTriviaList } from "../../components/PendingTriviaList";
import { useEffect, useState } from "react";
import * as TriviaService from "../../services/triviaService";
import { CreateTriviaBody, PendingTrivia } from "../../types/sharedTypes";
import { Footer } from "../../components/Footer";
import useAuth from "../../hooks/useAuth";
import { TRIVIA_CATEGORIES } from "../../constants/trivia";
import { CategoryCard } from "../../components/CategoryCard";

export default function Home() {
  const [pendingTrivia, setPendingTrivia] = useState<PendingTrivia[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const { userData } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    const fetchPendingTrivia = async () => {
      try {
        const { data } = await TriviaService.getPendingTrivia();
        setPendingTrivia(data.pendingTrivia);
      } catch (error) {
        console.log(error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchPendingTrivia();
  }, []);

  const handleCategoryCardClick = async (params: CreateTriviaBody) => {
    try {
      if (!userData) {
        localStorage.setItem("triviaConfig", JSON.stringify(params));
        navigate("/login");
        return;
      }

      const { data } = await TriviaService.createTrivia(params);

      navigate(`/trivia/${data.triviaId}`);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Header />

      <Box as="main">
        <Box width="100%" backgroundColor="purple.500" overflow="hidden">
          <Container
            padding={{ base: "40px 20px", md: "80px 0" }}
            position="relative"
          >
            <Text
              fontSize={{ base: "3xl", md: "6xl" }}
              lineHeight="1.2"
              fontWeight="bold"
              color="white"
              maxWidth="550"
            >
              Challenge yourself in a trivia match!
            </Text>

            <Button
              backgroundColor="green.400"
              color="white"
              width="180px"
              _hover="green.400"
              marginTop="38px"
              as={Link}
              to="/trivia/config"
            >
              Start a trivia
            </Button>
            <Image
              src="/assets/home.svg"
              alt="Home image"
              position="absolute"
              right="-160px"
              bottom="-60px"
              width="674px"
              height="418px"
              pointerEvents="none"
              display={{ base: "none", md: "none", lg: "block" }}
            />
          </Container>
        </Box>

        {userData && (
          <Container padding={{ base: "40px 20px", md: "80px 0 120px" }}>
            <Box as="header" marginBottom={{ base: "20px", md: "40px" }}>
              <Heading fontSize="3xl">Pending trivias</Heading>
            </Box>

            {isLoading ? (
              <Stack
                display="grid"
                gap="16px"
                gridTemplateColumns={{
                  base: "repeat(1fr)",
                  md: "repeat(4, 1fr)",
                }}
              >
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
              </Stack>
            ) : pendingTrivia.length ? (
              <PendingTriviaList pendingTriviaList={pendingTrivia} />
            ) : (
              <Text>No pending trivias</Text>
            )}
          </Container>
        )}

        <Box
          as="section"
          padding={{ base: "40px 20px", md: "80px 0 120px" }}
          backgroundColor="gray.100"
          boxShadow="0px 30px 100px 0 rgba(39, 39, 39, 0.2);"
        >
          <Box
            as="header"
            textAlign="center"
            marginBottom={{ base: "20px", md: "60px" }}
          >
            <Heading fontSize={{ base: "3xl", md: "5xl" }} marginBottom="20px">
              Show off your knowledge
            </Heading>
            <Text fontSize={{ base: "md", md: "xl" }}>
              Take a quiz and impress your friends with what you know!
            </Text>
          </Box>

          <Container
            display="grid"
            gridTemplateColumns={{ base: "repeat(1fr)", md: "repeat(5, 1fr)" }}
            gap="16px"
            padding={{ base: "0", md: "0 0" }}
          >
            {Object.values(TRIVIA_CATEGORIES).map((category) => (
              <CategoryCard
                key={category.title}
                title={category.title}
                icon={category.icon}
                color={category.color}
                onCardClick={() => handleCategoryCardClick(category.params)}
              />
            ))}
          </Container>
        </Box>
      </Box>

      <Footer />
    </>
  );
}
