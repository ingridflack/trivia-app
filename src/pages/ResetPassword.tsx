import { Box, Card, Container, Image, Text } from "@chakra-ui/react";
import Header from "../components/Header";
import ResetPasswordForm from "../components/ResetPasswordForm";

export default function ResetPassword() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding="70px"
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Reset your password
            </Text>
            <ResetPasswordForm />
          </Box>

          <Image
            src="/src/assets/recovery.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
          />
        </Card>
      </Container>
    </>
  );
}
