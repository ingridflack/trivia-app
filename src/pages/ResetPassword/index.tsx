import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import Header from "../../components/Header";
import ResetPasswordForm from "../../components/ResetPasswordForm";
import { useNavigate, useParams } from "react-router-dom";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../../services/authService";
import { Footer } from "../../components/Footer";

export interface ResetPasswordFormValues {
  password: string;
  passwordConfirmation: string;
}

export default function ResetPassword() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<ResetPasswordFormValues>();
  const toast = useToast();
  const navigate = useNavigate();
  const { token } = useParams();

  const onSubmit: SubmitHandler<ResetPasswordFormValues> = async (values) => {
    if (!token) return;

    try {
      await AuthService.resetPassword(token, values);

      toast({
        title: "Email sent.",
        description: "Check your inbox.",
        status: "success",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });

      navigate("/login");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };
  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding={{ base: "40px 20px", md: "80px 0" }}
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Reset your password
            </Text>
            <ResetPasswordForm
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              register={register}
              isSubmitting={isSubmitting}
            />
          </Box>

          <Image
            src="/assets/recovery.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
        </Card>
      </Container>
      <Footer />
    </>
  );
}
