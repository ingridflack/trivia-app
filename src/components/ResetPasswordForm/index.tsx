import {
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  Stack,
} from "@chakra-ui/react";
import { FieldErrors, UseFormRegister } from "react-hook-form";
import { ResetPasswordFormValues } from "../../pages/ResetPassword";

interface ResetPasswordFormProps {
  onSubmit: (e?: React.BaseSyntheticEvent) => Promise<void>;
  errors: FieldErrors<ResetPasswordFormValues>;
  register: UseFormRegister<ResetPasswordFormValues>;
  isSubmitting: boolean;
  token?: string;
}

export default function ResetPasswordForm({
  onSubmit,
  errors,
  register,
  isSubmitting,
  token,
}: ResetPasswordFormProps) {
  if (!token) {
    return (
      <div>
        <p>Invalid token.</p>
      </div>
    );
  }

  return (
    <Stack
      as="form"
      onSubmit={onSubmit}
      gap="16px"
      maxWidth="400px"
      role="form"
    >
      <FormControl isInvalid={!!errors.password}>
        <FormLabel htmlFor="password">Password</FormLabel>
        <Input
          type="password"
          placeholder="Password"
          id="password"
          {...register("password", {
            required: "Password is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
          })}
        />
        <FormErrorMessage>
          {errors.password && errors.password.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.passwordConfirmation}>
        <FormLabel htmlFor="passwordConfirmation">
          Password confirmation
        </FormLabel>
        <Input
          type="password"
          placeholder="Password Confirmation"
          id="passwordConfirmation"
          {...register("passwordConfirmation", {
            required: "Password confirmation is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
            validate: {
              isEqual: (value, { password }) =>
                value !== password ? "The passwords do not match" : true,
            },
          })}
        />
        <FormErrorMessage>
          {errors.passwordConfirmation && errors.passwordConfirmation.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        colorScheme="purple"
        size="lg"
        width="100%"
        isLoading={isSubmitting}
        type="submit"
      >
        Reset
      </Button>
    </Stack>
  );
}
