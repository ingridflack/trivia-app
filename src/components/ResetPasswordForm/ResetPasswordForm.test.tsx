import { FieldErrors } from "react-hook-form";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { afterEach, describe, expect, it, vi } from "vitest";
import ResetPasswordForm, { ResetPasswordFormProps } from ".";
import { ResetPasswordFormValues } from "../../pages/ResetPassword";

const setup = (props: Partial<ResetPasswordFormProps> = {}) =>
  render(
    <ResetPasswordForm
      onSubmit={vi.fn()}
      errors={{}}
      register={vi.fn()}
      isSubmitting={false}
      token="valid-token"
      {...props}
    />
  );

const formFieldsData = [
  {
    errors: { password: { message: "Invalid password" } },
    name: "password",
    regex: /invalid password/i,
  },
  {
    errors: {
      passwordConfirmation: { message: "Invalid password confirmation" },
    },
    name: "passwordConfirmation",
    regex: /invalid password confirmation/i,
  },
];

describe("ResetPasswordForm", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match the snapshot", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when the token is invalid", () => {
    const { container } = setup({ token: "" });
    expect(container).toMatchSnapshot();
  });

  it("should call onSubmit when the form is submitted", () => {
    const mockOnSubmit = vi.fn();
    const { getByRole } = setup({ onSubmit: mockOnSubmit });

    fireEvent.submit(getByRole("form"));
    expect(mockOnSubmit).toHaveBeenCalled();
  });

  it("should render the button in loading state when isSubmitting is true", () => {
    const { getByRole } = setup({ isSubmitting: true });
    const button = getByRole("button", { name: /reset/i });
    expect(button).toHaveProperty("disabled", true);
  });

  formFieldsData.forEach(({ errors, name, regex }) => {
    it(`should display the ${name} error message when it's invalid`, () => {
      const props = { errors: errors as FieldErrors<ResetPasswordFormValues> };
      const { getByText } = setup(props);
      expect(getByText(regex)).toBeTruthy();
    });
  });
});
