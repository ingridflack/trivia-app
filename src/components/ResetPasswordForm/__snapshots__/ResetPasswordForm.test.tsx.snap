// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResetPasswordForm > should match the snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-2umoaz"
    role="form"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="password"
        id="field-:r0:-label"
      >
        Password
      </label>
      <input
        class="chakra-input css-0"
        id="password"
        placeholder="Password"
        type="password"
      />
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="passwordConfirmation"
        id="field-:r1:-label"
      >
        Password confirmation
      </label>
      <input
        class="chakra-input css-0"
        id="passwordConfirmation"
        placeholder="Password Confirmation"
        type="password"
      />
    </div>
    <button
      class="chakra-button css-uxt1e8"
      type="submit"
    >
      Reset
    </button>
  </form>
</div>
`;

exports[`ResetPasswordForm > should match the snapshot when the token is invalid 1`] = `
<div>
  <div>
    <p>
      Invalid token.
    </p>
  </div>
</div>
`;
